src/hev-ifwd.c

```diff
@@ -0,0 +1,451 @@
+/*
+ ============================================================================
+ Name        : hev-ifwd.c
+ Author      : Mike Wang <mikewang000000@gmail.com>
+               hev <r@hev.cc>
+ Copyright   : Copyright (c) 2024 xyz
+ Description : iptables forwarder
+ ============================================================================
+ */
+
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#include "hev-conf.h"
+#include "hev-ifwd.h"
+#include "hev-misc.h"
+#include "hev-exec.h"
+
+#define BUFFLEN 1024
+
+typedef struct _IptNatRule IptNatRule;
+
+struct _IptNatRule
+{
+    char table[16];
+    char chain[32];
+    char proto[8];
+    char ipv4[16];
+    char port[16];
+    char toipv4[16];
+    char toport[16];
+};
+
+static int ipt_initialized = 0;
+static int ipt_atexit      = 0;
+static IptNatRule nat_rule;
+
+static int
+call_subprocess (const char *file, char *const argv[], char *sstdout,
+                 size_t nstdout, char *sstderr, size_t nstderr)
+{
+    int pipefd_o[2], pipefd_e[2];
+    char blackhole[512];
+    int status = -1;
+    int nbyte;
+    pid_t pid;
+
+    signal(SIGCHLD, SIG_DFL);
+
+    if (pipe (pipefd_o) == -1 || pipe (pipefd_e) == -1) {
+        LOGV (E, "%s", strerror (errno));
+        goto err_exit;
+    }
+
+    pid = fork ();
+
+    if (pid == -1) {
+        LOGV (E, "%s", strerror (errno));
+        goto err_close;
+    }
+
+    if (pid == 0) {
+        if (dup2 (pipefd_o[1], STDOUT_FILENO) == -1 ||
+            dup2 (pipefd_e[1], STDERR_FILENO) == -1) {
+            LOGV (E, "%s", strerror (errno));
+            _exit (EXIT_FAILURE);
+        }
+        close (pipefd_o[0]);
+        close (pipefd_o[1]);
+        close (pipefd_e[0]);
+        close (pipefd_e[1]);
+        execvp (file, argv);
+        LOGV (E, "%s", strerror (errno));
+        _exit (EXIT_FAILURE);
+    }
+
+    close (pipefd_o[1]);
+    close (pipefd_e[1]);
+
+    nbyte = read (pipefd_o[0], sstdout, nstdout ? nstdout - 1 : 0);
+    if (nbyte == -1) {
+        LOGV (E, "%s", strerror (errno));
+        goto err_close_read;
+    }
+
+    if (nstdout)
+        sstdout[nbyte] = 0;
+
+    while (read (pipefd_o[0], blackhole, sizeof (blackhole)) > 0) { }
+
+    nbyte = read (pipefd_e[0], sstderr, nstderr ? nstderr - 1 : 0);
+    if (nbyte == -1) {
+        LOGV (E, "%s", strerror (errno));
+        goto err_close_read;
+    }
+
+    if (nstderr)
+        sstderr[nbyte] = 0;
+
+    while (read (pipefd_e[0], blackhole, sizeof (blackhole)) > 0) { }
+
+    if (waitpid (pid, &status, 0) == -1) {
+        LOGV (E, "%s", strerror (errno));
+        goto err_close_read;
+    }
```

Comment:
`waitpid()` here is somehow conflict with https://github.com/heiher/natmap/blob/c564210613e1198059ebd547d37a782b2fd27f6e/src/hev-exec.c#L46